File: native/bindings/endstone/inventory/item_type.cpp
#include "item_type.h"

#include "item_stack.h"

#include <endstone/inventory/item_type.h>
#include <endstone/server.h>

#include <memory>
#include <string>
#include <string_view>

namespace aegilex::native::inventory {

class ItemType::impl {
  public:
    explicit impl(const endstone::ItemType *type) noexcept : type(type)
    {
    }

    const endstone::ItemType *type;
};

ItemType::ItemType(const endstone::ItemType *type) noexcept : impl(std::make_shared<class ItemType::impl>(type))
{
}

const endstone::ItemType *ItemType::native() const noexcept
{
    return impl == nullptr ? nullptr : impl->type;
}

rust::String ItemType::getTypeId() const
{
    try {
        return rust::String(std::string(impl->type->getId()));
    }
    catch (...) {
        return rust::String();
    }
}

rust::String ItemType::getTranslationKey() const
{
    try {
        return rust::String(impl->type->getTranslationKey());
    }
    catch (...) {
        return rust::String();
    }
}

std::int32_t ItemType::getMaxStackSize() const
{
    try {
        return impl->type->getMaxStackSize();
    }
    catch (...) {
        return 0;
    }
}

std::int32_t ItemType::getMaxDurability() const
{
    try {
        return impl->type->getMaxDurability();
    }
    catch (...) {
        return 0;
    }
}

std::unique_ptr<ItemStack> ItemType::createItemStack(const std::int32_t amount) const
{
    try {
        if (amount < 1 || amount > impl->type->getMaxStackSize()) {
            return std::unique_ptr<ItemStack>();
        }
        return std::unique_ptr<ItemStack>(new ItemStack(impl->type->createItemStack(amount)));
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

} // namespace aegilex::native::inventory

File: rust/crates/aegilex-runtime/src/host/endstone/inventory/item_type.rs
//! Core ABI implementation for `native/bindings/endstone/inventory/item_type.h`.

use crate::host::endstone::support::*;

impl crate::core_host::imports::HostItemType for PluginStoreState {
    fn get_item_type(
        &mut self,
        type_id: String,
    ) -> Result<Result<ItemTypeItemTypeData, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "item-type.get-item-type")?;
            let item_type = resolve_server(self)
                .map_err(map_core_host_error)?
                .getItemType(&type_id);
            if item_type.is_null() {
                return Err(map_core_host_error(not_found()));
            }
            Ok(ItemTypeItemTypeData {
                type_id: item_type.getTypeId(),
                translation_key: item_type.getTranslationKey(),
                max_stack_size: item_type.getMaxStackSize(),
                max_durability: item_type.getMaxDurability(),
            })
        })())
    }

    fn create_item_stack(
        &mut self,
        type_id: String,
        amount: Option<i32>,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "item-type.create-item-stack")?;
            let item_type = resolve_server(self)
                .map_err(map_core_host_error)?
                .getItemType(&type_id);
            if item_type.is_null() {
                return Err(map_core_host_error(not_found()));
            }
            let item = item_type.createItemStack(amount.unwrap_or(1));
            if item.is_null() {
                return Err(map_core_host_error(not_found()));
            }
            self.insert_item_stack_resource(item)
                .map_err(map_core_host_error)
        })())
    }
}
